### ISSUE TITLE
[ci] Should improve bazel caching

### ISSUE THREAD

**Issue by EricCousineau-TRI** [RobotLocomotion/drake-ros#247]
The past few PRs have been very slow to build / use `drake@v1.13.0` per recent update.

May possible be improved by following up on https://drakedevelopers.slack.com/archives/C2CP2E50A/p1673880604846249

**Comment by jwnimmer-tri** [comment]
I'm pretty sure that the caching is all wrong.

By default, GitHub Actions only _write back_ to the cache when the `key:` string changes.  Since our key string is a literal with no substitutions, it is never written back, so it's getting progressively colder as we continue to change the code.

See https://github.com/RobotLocomotion/drake-blender/pull/7 where I'm iterating to get this right.  I'll circle back here once I've tested it there on a smaller repo.

**Comment by EricCousineau-TRI** [comment]
Ah, good to know, and thanks!

**Comment by EricCousineau-TRI** [comment]
Getting close, but the job can still get cancelled sometimes:
https://github.com/RobotLocomotion/drake-ros/actions/runs/6303897969/job/17114097008

It seems like it is still running out of disk space:
https://github.com/RobotLocomotion/drake-ros/actions/runs/6303897969/attempts/1
```
Unhandled exception. System.IO.IOException: No space left on device : '/home/runner/runners/2.309.0/_diag/Worker_20230925-192117-utc.log'
```
@adityapande-1995 Would you be able to look into this?
\cc @calderpg-tri @IanTheEngineer 

FWIW, GitHub Workflows seems a bit wonky right now:
- I don't always see instructions for the current executing step
- If I re-run the job, then I cannot see the old logs

**Comment by jwnimmer-tri** [comment]
BTW If you're running `apt` commands in the build (either update and/or upgrade), make sure you nuke their downloaded files and lists database afterwards:
https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#apt-get

**Comment by EricCousineau-TRI** [comment]
Resolved by #309
